[UnifiedPDF][iOS] Relative scroll position is not preserved in zoomed content, following a viewport update

pxlcoder opened the pull request:
#### c48b2af7092c5e7b62355010f2c66e704fc64a7f
<pre>
[UnifiedPDF][iOS] Relative scroll position is not preserved in zoomed content, following a viewport update
<a href="https://bugs.webkit.org/show_bug.cgi?id=294472">https://bugs.webkit.org/show_bug.cgi?id=294472</a>
<a href="https://rdar.apple.com/153191283">rdar://153191283</a>

Reviewed by Wenson Hsieh and Abrar Rahman Protyasha.

On iOS, the contents size of the frame view is changed to &quot;fit&quot; a displayed PDF.
In general, when the contents size of the frame view changes, `ScrollView::setContentsSize`
may update the scroll position underneath `ScrollView::updateScrollbars`.

`updateScrollbars` adjusts the desired position to fit with the new minimum and
maximum positions, and requests a programmatic scroll if the adjusted position
is different from the current position. When zoomed and scrolled, a viewport
update may reduce the maximum position, resulting in the x-coordinate of the
desired position getting clamped. At the time of the content size update in the
web process, the y-coordinate is still the same value as it was before the
content size update.

Consequently, the adjusted position differs in the x-coordinate, and a
programmatic scroll request is made. Before the request is serviced, the logic
added in 295715@main kicks in and attempts to preserve the relative scroll
position. However, the scroll request is then serviced, pulling the y-coordinate
back to the old value, negating the fix in 295715@main.

To fix, simply prohibit scrolling due to content size changes in the web process.
The scroll position is already corrected in the UI process for PDFs, making the
other logic unnecessary and risky.

* Source/WebCore/platform/ScrollView.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:

Use an existing mechanism to prohibit scrolling due to content size changes.

* Tools/TestWebKitAPI/Tests/WebKitCocoa/UnifiedPDFTests.mm:
(TestWebKitAPI::UNIFIED_PDF_TEST(KeepRelativeScrollPositionAfterZoomingAndViewportUpdate)):

Swizzle `-[UIPinchGestureRecognizer state]` to simulate a user pinch zooming.
This is necessary, as otherwise the initial scale of 1 is restored when the
layout size is changed programmatically.

Canonical link: <a href="https://commits.webkit.org/296222@main">https://commits.webkit.org/296222@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/870d1d7b2acc7ae11da43c56499a0b5793eabc54

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/107797 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/27467 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/17884 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/113013 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/58323 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/109763 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/28162 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/36008 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/81828 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/110742 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/22313 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/97130 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/62242 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/21720 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/15270 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/57766 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/91673 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/15317 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/116134 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/34868 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/25676 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/90861 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/35244 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/93393 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/90645 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/35539 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/13307 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/30618 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/17429 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/34772 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/40324 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/34513 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/37875 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/36176 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aprotyas reviewed (commented):
Small typo in the commit message?

```
scroll position is already correcting the UI process
```

aprotyas commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/UnifiedPDFTests.mm:
```diff
@@ -550,6 +555,43 @@ void pressKey(auto key, unsigned short code, Seconds duration = 200_ms)
+            [UIPinchGestureRecognizer class],
+            @selector(state),
+            imp_implementationWithBlock(^UIGestureRecognizerState {
+                return UIGestureRecognizerStateBegan;
```
Nice

jesxilin commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -198,6 +198,11 @@ - (void)formChanged:(NSNotification *)notification
     if (m_presentationController->wantsWheelEvents())
         wantsWheelEventsChanged();
+
+    if (shouldSizeToFitContent()) {
```
Is it necessary/purposeful for this to be a nested if stmt?

pxlcoder replied:
What would the alternative be?

whsieh replied:
I suppose something like this might work?

```
if (RefPtr frameView = m_frame->coreLocalFrame()->view(); frameView && shouldSizeToFitContent())
    m_prohibitScrollingDueToContentSizeChanges = frameView->prohibitScrollingWhenChangingContentSizeForScope();
```

(…but it's unclear whether that's really cleaner)

pxlcoder committed:
[UnifiedPDF][iOS] Relative scroll position is not preserved in zoomed content, following a viewport update
https://bugs.webkit.org/show_bug.cgi?id=294472
rdar://153191283

Reviewed by Wenson Hsieh and Abrar Rahman Protyasha.

On iOS, the contents size of the frame view is changed to "fit" a displayed PDF.
In general, when the contents size of the frame view changes, `ScrollView::setContentsSize`
may update the scroll position underneath `ScrollView::updateScrollbars`.

`updateScrollbars` adjusts the desired position to fit with the new minimum and
maximum positions, and requests a programmatic scroll if the adjusted position
is different from the current position. When zoomed and scrolled, a viewport
update may reduce the maximum position, resulting in the x-coordinate of the
desired position getting clamped. At the time of the content size update in the
web process, the y-coordinate is still the same value as it was before the
content size update.

Consequently, the adjusted position differs in the x-coordinate, and a
programmatic scroll request is made. Before the request is serviced, the logic
added in 295715@main kicks in and attempts to preserve the relative scroll
position. However, the scroll request is then serviced, pulling the y-coordinate
back to the old value, negating the fix in 295715@main.

To fix, simply prohibit scrolling due to content size changes in the web process.
The scroll position is already corrected in the UI process for PDFs, making the
other logic unnecessary and risky.

* Source/WebCore/platform/ScrollView.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:

Use an existing mechanism to prohibit scrolling due to content size changes.

* Tools/TestWebKitAPI/Tests/WebKitCocoa/UnifiedPDFTests.mm:
(TestWebKitAPI::UNIFIED_PDF_TEST(KeepRelativeScrollPositionAfterZoomingAndViewportUpdate)):

Swizzle `-[UIPinchGestureRecognizer state]` to simulate a user pinch zooming.
This is necessary, as otherwise the initial scale of 1 is restored when the
layout size is changed programmatically.

Canonical link: https://commits.webkit.org/296222@main
